streams_map.go

```diff
@@ -69,14 +71,42 @@ func (m *streamsMap) GetOrOpenStream(id protocol.StreamID) (*stream, error) {
 		return s, nil // s may be nil
 	}
 
-	// ... we don't have an existing stream, try opening a new one
+	// ... we don't have an existing stream
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 	// We need to check whether another invocation has already created a stream (between RUnlock() and Lock()).
 	s, ok = m.streams[id]
 	if ok {
 		return s, nil
 	}
+
+	if id <= m.highestStreamOpenedByPeer {
+		return nil, nil
+	}
+
+	highestOpened := m.highestStreamOpenedByPeer
+	sid := id
+	for sid > highestOpened {
+		_, err := m.openRemoteStream(sid)
+		if err != nil {
+			return nil, err
+		}
+		if sid == 1 {
+			break
+		}
+		sid -= 2
```

Comment:
Maybe add a comment that highestStreamOpenedByPeer (and therefore sid) is always odd.